### PR TITLE
Add GitProvider to retrieve conf file from Git Server.

### PR DESCRIPTION
**Pull request by felipebonezi**
# Version 1.0.0

- ✨ **Add GitProvider to retrieve conf file from Git Server.**
   - You can retrieve file using three auth modes:
        1. None or Empty (For public repositories);
        2. user (For private repositories over login and password);
        3. ssh-rsa (For private repositories over SSH RSA private key).

**Issue Ref:** [#1](https://github.com/play-rconf/play-rconf-sdk/issues/1)

### CODE REVIEW

**Comment by felipebonezi** [comment]
@thibaultmeyer how do you feel about this code?
I'll implement the SSH RSA over FILE PATH and FILE CONTENT in the next commit.

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -0,0 +1,37 @@
+remote-configuration {
+
+  ## Provider - Git
+  # ~~~~~
+  # Retrieves configuration from a simple Git repository
+  http {
```
Comment (thibaultmeyer) [inline src/main/resources/reference.conf]: Rename it as `git`

File: src/main/java/io/playrconf/provider/GitProvider.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    /**
+     * Clones the branch of Git repository.
+     *
+     * @param config Config file.
+     * @param repositoryURI Repository URI using HTTPS or SSH.
+     * @param mode Auth mode.
+     * @param branch Branch name.
+     * @return Repository.
+     */
+    private Repository cloneRepository(Config config, String repositoryURI, String mode, String branch) throws GitAPIException {
```
Comment (thibaultmeyer) [inline src/main/java/io/playrconf/provider/GitProvider.java]: Please, make method parameters `final`

https://travis-ci.org/github/play-rconf/play-rconf-git/builds/740187581
```
[ERROR] GitProvider.java:115:26: Parameter config should be final. [FinalParameters]
[ERROR] GitProvider.java:116:26: Parameter kvObjConsumer should be final. [FinalParameters]
[ERROR] GitProvider.java:117:26: Parameter fileObjConsumer should be final. [FinalParameters]
[ERROR] GitProvider.java:121:20: Variable 'mode' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:122:20: Variable 'repositoryURI' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:123:20: Variable 'branch' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:124:20: Variable 'filepath' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:126:24: Variable 'repository' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:127:22: Variable 'head' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:128:23: Variable 'lastCommit' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:130:20: Variable 'conf' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:178:44: Parameter config should be final. [FinalParameters]
[ERROR] GitProvider.java:191:16: Variable 'mode' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:192:16: Variable 'repositoryURI' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:224:40: Parameter config should be final. [FinalParameters]
[ERROR] GitProvider.java:224:55: Parameter repositoryURI should be final. [FinalParameters]
[ERROR] GitProvider.java:224:77: Parameter mode should be final. [FinalParameters]
[ERROR] GitProvider.java:224:90: Parameter branch should be final. [FinalParameters]
[ERROR] GitProvider.java:225:14: Variable 'repoDir' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:226:22: Variable 'cloneCommand' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:232:35: Variable 'sshSessionFactory' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:235:54: Parameter fs should be final. [FinalParameters]
[ERROR] GitProvider.java:237:32: Variable 'privateKey' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:238:32: Variable 'password' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:238:78: '?' should be on a new line. [OperatorWrap]
[ERROR] GitProvider.java:240:30: Variable 'jSch' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:247:34: Variable 'sshTransport' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:253:24: Variable 'username' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:254:24: Variable 'password' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:257:13: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] GitProvider.java:272:29: Parameter repository should be final. [FinalParameters]
[ERROR] GitProvider.java:272:52: Parameter commit should be final. [FinalParameters]
[ERROR] GitProvider.java:272:70: Parameter filepath should be final. [FinalParameters]
[ERROR] GitProvider.java:273:18: Variable 'walk' should be declared final. [FinalLocalVariable]
[ERROR] GitProvider.java:278:16: Variable 'bytes' should be declared final. [FinalLocalVariable]
```

File: src/main/java/io/playrconf/provider/GitProvider.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        File repoDir = new File("play-rconf-git", String.format("%s-%s-%d", repositoryURI, branch, System.currentTimeMillis()));
```
Comment (thibaultmeyer) [inline src/main/java/io/playrconf/provider/GitProvider.java]: Is the directory deleted after ?

**Review by thibaultmeyer — CHANGES_REQUESTED** [review]
There are just a few small changes to be made.

File: src/test/java/io/playrconf/provider/GitProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+    public void gitTest_010() {
+        loadConfigWithError(INITIAL_CONFIGURATION_ERROR_MISSING_PASSWORD_FOR_USER_MODE);
+    }
+
+    @Test(expected = ConfigException.Missing.class)
+    public void gitTest_011() {
+        loadConfigWithError(INITIAL_CONFIGURATION_ERROR_MISSING_PRIVATE_KEY_FOR_SSH_MODE);
+    }
+
+    @Test(expected = ConfigException.BadValue.class)
```
Comment (thibaultmeyer) [inline src/test/java/io/playrconf/provider/GitProviderTest.java]: `ConfigException.BadPath.class`

**Review by thibaultmeyer — CHANGES_REQUESTED** [review]

**Comment by thibaultmeyer** [comment]
@felipebonezi
I have added RSA keys. 1 with password (password = travis) and 1 without password. The files are located into `.github` directory and will be used by TravisCI

**Comment by felipebonezi** [comment]
> @felipebonezi
> I have added RSA keys. 1 with password (password = travis) and 1 without password. The files are located into `.github` directory and will be used by TravisCI

@thibaultmeyer I think you generated the wrong format private key.
You need to generate a RSA PRIVATE KEY and it's using OPENSSH PRIVATE KEY as you can see.

Captura de Tela 2020-10-31 às 17 00 51

Read more how to create a RSA Key: 
https://www.scottbrady91.com/OpenSSL/Creating-RSA-Keys-using-OpenSSL

**Comment by thibaultmeyer** [comment]
hmmm , Github only works with SSH key (deployment key with Read Only permission). You can see the documentation https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/connecting-to-github-with-ssh

I'm not sure we can run unit tests for the client side authentication via SSL certificat on github/travisCI. Maybe you can remove these related unit tests to pass the continuous integration check

File: src/main/java/io/playrconf/provider/GitProvider.java (hunk 3 above)
Comment (felipebonezi) [reply src/main/java/io/playrconf/provider/GitProvider.java]: @thibaultmeyer I've implemented the `Temp Folder` over `Files.createTempDirectory(...)` method.

**Comment by felipebonezi** [comment]
@thibaultmeyer all checks have passed ✅

**Review by thibaultmeyer — APPROVED** [review]

**Comment by thibaultmeyer** [comment]
Thank you

**Pull request merged**